data/transactions/logic/eval.go

```diff
@@ -1148,3 +1537,541 @@ func opStore(cx *evalContext) {
+		cx.err = errors.New("concat resulted in string too long")
+		return
+	}
+	newvalue := make([]byte, newlen)
+	copy(newvalue, a)
+	copy(newvalue[len(a):], b)
+	cx.stack[prev].Bytes = newvalue
+	cx.stack = cx.stack[:last]
+}
+
+func substring(x []byte, start, end int) (out []byte, err error) {
+	out = x
+	if end <= start {
+		err = errors.New("substring end before start")
+		return
+	}
+	if start > len(x) || end > len(x) {
+		err = errors.New("substring range beyond length of string")
+		return
+	}
+	out = x[start:end]
+	err = nil
+	return
+}
+
+func opSubstring(cx *evalContext) {
+	last := len(cx.stack) - 1
+	start := cx.program[cx.pc+1]
+	end := cx.program[cx.pc+2]
+	cx.stack[last].Bytes, cx.err = substring(cx.stack[last].Bytes, int(start), int(end))
+	cx.nextpc = cx.pc + 3
+}
+
+const maxInt = int((^uint(0)) >> 1)
+
+func opSubstring3(cx *evalContext) {
+	last := len(cx.stack) - 1 // end
+	prev := last - 1          // start
+	pprev := prev - 1         // bytes
+	start := cx.stack[prev].Uint
+	end := cx.stack[last].Uint
+	if start > uint64(maxInt) || end > uint64(maxInt) {
+		cx.err = errors.New("substring range beyond length of string")
+		return
+	}
+	cx.stack[pprev].Bytes, cx.err = substring(cx.stack[pprev].Bytes, int(start), int(end))
+	cx.stack = cx.stack[:prev]
+}
+
+func opBalance(cx *evalContext) {
+	last := len(cx.stack) - 1 // account offset
+
+	accountIdx := cx.stack[last].Uint
+
+	if cx.Ledger == nil {
+		cx.err = fmt.Errorf("ledger not available")
+		return
+	}
+
+	addr, err := cx.Txn.Txn.AddressByIndex(accountIdx, cx.Txn.Txn.Sender)
+	if err != nil {
+		cx.err = err
+		return
+	}
+
+	microAlgos, err := cx.Ledger.Balance(addr)
+	if err != nil {
+		cx.err = fmt.Errorf("failed to fetch balance of %s: %s", addr, err.Error())
+		return
+	}
+
+	cx.stack[last].Uint = microAlgos.Raw
+}
+
+func opAppCheckOptedIn(cx *evalContext) {
+	last := len(cx.stack) - 1 // app id
+	prev := last - 1          // account offset
+
+	appID := cx.stack[last].Uint
+	accountIdx := cx.stack[prev].Uint
+
+	if cx.Ledger == nil {
+		cx.err = fmt.Errorf("ledger not available")
+		return
+	}
+
+	addr, err := cx.Txn.Txn.AddressByIndex(accountIdx, cx.Txn.Txn.Sender)
+	if err != nil {
+		cx.err = err
+		return
+	}
+
+	_, err = cx.Ledger.AppLocalState(addr, basics.AppIndex(appID))
+	if err != nil {
+		cx.stack[prev].Uint = 0
+	} else {
+		cx.stack[prev].Uint = 1
+	}
+
+	cx.stack = cx.stack[:last]
+}
+
+func (cx *evalContext) getReadOnlyLocalState(appID uint64, accountIdx uint64) (basics.TealKeyValue, error) {
+	// Convert the account offset to an address
+	addr, err := cx.Txn.Txn.AddressByIndex(accountIdx, cx.Txn.Txn.Sender)
+	if err != nil {
+		return nil, err
+	}
+
+	kvIdx := ckey{appID, addr}
+	localKV, ok := cx.readOnlyLocalStates[kvIdx]
+	if !ok {
+		var err error
+		localKV, err = cx.Ledger.AppLocalState(addr, basics.AppIndex(appID))
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch app local state for acct %s, app %d: %v", addr, appID, err)
```

Comment:
If the `addr` is of type `basics.Address`, could you use `%v` instead of `%s` when formatting it ?

Reply:
sure, done :+1: